src/core/net/tcp6.cpp

```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
+    tcp.mEndpoints.PopAfter(prev);
+    SetNext(nullptr);
+
+exit:
+    return error;
+}
+
+uint8_t Tcp::Endpoint::TimerFlagToIndex(uint8_t aTimerFlag)
+{
+    OT_UNUSED_VARIABLE(aTimerFlag);
+    /*
+     * TODO: Convert from the timer flag provided by TCPlp to the index in
+     * our timers array.
+     */
+    return 0;
+}
+
+bool Tcp::Endpoint::IsTimerActive(uint8_t aTimerIndex)
+{
+    OT_UNUSED_VARIABLE(aTimerIndex);
+    /* TODO: Check whether TCPlp has marked this timer as active. */
+    return false;
+}
+
+void Tcp::Endpoint::SetTimer(uint8_t aTimerFlag, uint32_t aDelay)
+{
+    TimeMilli now         = TimerMilli::GetNow();
+    TimeMilli newFireTime = now + aDelay;
+    uint8_t   timerIndex  = TimerFlagToIndex(aTimerFlag);
+
+    mTimers[timerIndex] = newFireTime.GetValue();
+    otLogDebgTcp("Endpoint %p set timer %u with delay %u", this, static_cast<unsigned int>(timerIndex),
+                 static_cast<unsigned int>(aDelay));
+
+    GetInstance().Get<Tcp>().ResetTimer(false);
+}
+
+void Tcp::Endpoint::CancelTimer(uint8_t aTimerFlag)
+{
+    uint8_t timerIndex = TimerFlagToIndex(aTimerFlag);
+
+    /*
+     * TCPlp has already cleared the timer flag before calling this, so all
+     * that we have to do is reset the main timer if needed.
+     */
+
+    otLogDebgTcp("Endpoint %p cancelled timer %u", this, static_cast<unsigned int>(timerIndex));
+    if (GetInstance().Get<Tcp>().mTimer.GetFireTime() == TimeMilli(mTimers[timerIndex]))
+    {
+        GetInstance().Get<Tcp>().ResetTimer(false);
+    }
+}
+
+void Tcp::Endpoint::FirePendingTimers(TimeMilli aNow)
+{
+    /*
+     * NOTE: Firing a timer might potentially activate/deactivate other timers.
+     * If timers x and y expire at the same time, but the callback for timer x
+     * (for x < y) cancels or postpones timer y, should timer y's callback be
+     * called? Our answer is no, since timer x's callback has updated the
+     * TCP stack's state in such a way that it no longer expects timer y's
+     * callback to to be called. Because the TCP stack thinks that timer y
+     * has been cancelled, calling timer y's callback could potentially cause
+     * problems.
+     */
+    for (uint8_t timerIndex = 0; timerIndex != kNumTimers; timerIndex++)
+    {
+        if (IsTimerActive(timerIndex))
+        {
+            TimeMilli expiry(mTimers[timerIndex]);
+
+            if (expiry <= aNow)
+            {
+                /* TODO: Call TCPlp's callback for this timer. */
+            }
+        }
+    }
+}
+
+bool Tcp::Endpoint::GetEarliestActiveTimer(TimeMilli aNow, TimeMilli &aExpiryTime)
+{
+    bool hasUnfiredTimer = false;
+
+    aExpiryTime = aNow.GetDistantFuture();
+
+    for (uint8_t timerIndex = 0; timerIndex != kNumTimers; timerIndex++)
+    {
+        if (IsTimerActive(timerIndex))
+        {
+            TimeMilli expiry(mTimers[timerIndex]);
+
+            hasUnfiredTimer = true;
+            aExpiryTime     = OT_MIN(aExpiryTime, expiry);
```

Comment:
I think the current code is correct.

Even in the previous version, I think that it would have been safe. If you call `Timer::FireAt` with a time that's in the past (`< now` as compared in the timer code) then the alarm is set to 0 msec and the timer tasklet will be posted immediately. See the code here: https://github.com/openthread/openthread/blob/main/src/core/common/timer.cpp#L173. So there's no harm in calling `GetEarliestActiveTimer()` even if there are expired timers, since you will choose an expired timer, set  `mTimer` to fire at its expired time in the past, causing the alarm to be set to 0 msec, and then call those callbacks when the Timer callback occurs.

I did implement your suggestion for the reasons described in the comment above (simplicity and overhead of linked list manipulations vs extra wakeups). So I believe the current code is correct, regardless of my explanation in the previous paragraph.

Reply:
Yes. The `FireAt()` works correctly if the fire time is in the past (that's very much intentional in the timer design and we use and rely on it often 😃).

But that was not the place that could potentially cause an issue if we call `GetEarliestActiveTimer()` with expired timers in the previous code. The issue could happen on `OT_MIN()` when we try to keep track of the next earliest fire time.
- We init the `aExpiryTime = aNow.GetDistantFuture()`.
- By definition `aNow.GetDistantFuture()` gives a time that is ahead of any other time that is ahead of `aNow`. 
- If we now compare `aExpireTime` with a time that is past of `aNow` (expired timer) that won't work correctly.
- Note that the same situation can happen if we don't use `GetDistantFuture()` and instad of try to track the delay duration (i.e. if we try to get delay till next expire time using `(aExpiryTime - aNow)` we need to check and ensure it is ahead of `aNow`).

We need to either ensure expired times are handled  or update code to check for this case as we try to find next earliest fire time.

Reply:
OK, thanks for clarifying. I hadn't thought about that. I think the previous version probably had this bug, but the latest version doesn't have it. That's because there are no `ResetTimer(false)` calls in the latest version, and I believe that `GetEarlisetActiveTimer()` will never be called when there are pending timers.

Reply:
To make this even more clear, I actually combined `GetEarliestActiveTimer` and `FirePendingTimers` in the latest revision. I realized that, with `SetTimer` now calling `FireAtIfEarlier`, we no longer need to scan the list of timers in each TCP endpoint twice.

Reply:
Sounds good to me. Thanks. 😃